[7467] Add "Collapse/Expand" button to sidebar.

cjcenizal opened the pull request:
Addresses https://github.com/elastic/kibana/issues/7467.

Blocked by https://github.com/elastic/kibana/pull/7711. Once that PR is merged, we'll rebase this one.
## Changes
- Allow user to pin/unpin the sidebar, and store preference in localStorage.
- Add "Expand/Collapse" button to sidebar.
- Add appSwitcherState service.
- Add support for "classes" attribute to app-switcher-link.
- Remove animation from sidebar, and no longer expand on hover.
- Add default tooltips for sidebar items.
### Expanded and collapsed states

![sidebar_pin](https://cloud.githubusercontent.com/assets/1238659/16778895/e7acb6b0-4824-11e6-9f68-4499f44574dd.gif)

**Note:** If there is a tooltip specified for an app, then the app's title will be prepended to the tooltip text, e.g. "Visualize - Make your data beautiful."
### Tooltips on disabled links

![image](https://cloud.githubusercontent.com/assets/1238659/16778554/5d76f10a-4823-11e6-81b3-831920479586.png)

kjbekkelund commented:
Can you include a screenshot or gif in PRs such as this? Makes it a lot easier for us that don't have Kibana running all the time to see how it looks like (and to review the non-code parts of the PR)

(I recommend Gif Brewery if you want to make gifs, it's awesome)

cjcenizal commented:
@kjbekkelund Great call on Gif Brewery.

alt74 commented:
very nice and simple. I dig it. thanks @cjcenizal 👍

Bargs commented:
You added the tooltips ❤️ ❤️ ❤️ ❤️

kjbekkelund commented on src/ui/public/chrome/directives/app_switcher_link/__tests__/app-switcher-link.js:
```diff
@@ -207,5 +208,15 @@ describe('appSwitcherLink directive', () => {
+          appSwitcherLinkClasses: 'new classes',
+        };
+        const element = create(attrs);
+        expect(element.attr('class')).to.contain(attrs.appSwitcherLinkClasses);
```
I've never been a big fan of tests that use the same data as input and as what they check. What if `create(attrs)` didn't work so `class` is wrong, but because you have a typo in `attrs.appSwitcherLinkClassses` (aka it's `undefined`) it doesn't fail? I much prefer actually typing it out.

cjcenizal replied:
I think that's a very reasonable point. The reason I like this pattern is that it makes intention crystal-clear; there's no doubt what I want to happen in this test because the variable carries meaning. If I'm comparing against a literal value, then I need to read the entire test (or rely on a comment) to extract that meaning. Whenever I read a test and find it confusing, it's usually because I don't understand the intention behind the test.

I don't think I can objectively say this way is better than your suggestion, though, since there are pros and cons to both.

kjbekkelund commented:
LGTM

Bargs commented:
@cjcenizal I think calling the new service `appSwitcherState` is a little confusing since it affects more than just the app-switcher directive. Until I realized this controls the entire sidebar it seemed really odd that `kbn_chrome.js` had such intimate knowledge of the appSwitcherState". It seems like `chromeState` or `kbnChromeState` would be a more apt name, no?

Bargs commented on src/ui/public/chrome/chrome.html:
```diff
@@ -27,6 +30,15 @@
+
+      <!-- Open/close sidebar -->
+      <app-switcher-link
+        app-switcher-link-classes="appSwitcherButton.classes"
```
I don't think it's necessary to add this custom attribute to the app-switcher-link directive. You can change this line to `class="{{appSwitcherButton.classes}}"`

cjcenizal replied:
Haha, old instinct from writing React components. 😄

Bargs commented on src/ui/public/chrome/directives/app_switcher_link/app_switcher_link.less:
```diff
@@ -22,18 +23,19 @@
+
+  .is-app-links-wrapper-open & {
+    width: @as-open-width;
+  }
```
This is a pretty confusing use of a nested selector. Couldn't it just be a separate rule with a selector like `.is-app-links-wrapper-open .app-switcher-link`

cjcenizal replied:
Yes, it could... when I first saw this kind of "pattern", I also had a "woah, weird" reaction. After seeing it a few times I'm used to it now, and I also appreciate how it keeps the state grouped with the component.

I just tried pulling it out and seeing how it looked:

```less
.app-switcher-link {
  position: relative;
  width: @as-closed-width;
  height: @app-icon-height;
  line-height: @app-line-height;

  &.is-app-switcher-link-disabled {
    opacity: 0.5;

    .app-switcher-link__anchor {
      cursor: default;
    }
  }

  &:hover,
  &.active {
    background-color: @app-links-active-background;

    .app-switcher-link__anchor {
      color: @white;
      text-decoration: none;
    }
  }
}

.is-app-links-wrapper-open {
  .app-switcher-link {
    width: @as-open-width;
  }
}
```

It definitely becomes a little less readable to me. Mind if we keep it as-is?

cjcenizal replied:
Actually, after staring at it for a little longer, it's not that bad. I'll make the change.

The things I do for Bargs...

Bargs commented on src/ui/public/chrome/directives/kbn_chrome.js:
```diff
@@ -43,6 +43,29 @@ export default function (chrome, internals) {
+          event.preventDefault();
+          $scope.isAppSwitcherOpen = !$scope.isAppSwitcherOpen;
+          appSwitcherState.setOpen($scope.isAppSwitcherOpen);
+          updateAppSwitcher($scope.isAppSwitcherOpen);
```
If `appSwitcherState.setOpen` gets called outside of this `toggleAppSwitcher` function, `$scope.isAppSwitcherOpon` won't get updated. I'm always wary when I see scope variables and services that represent the same thing, but aren't linked via some kind of watcher that guarantees (as much as possible) they'll stay in sync.

Perhaps the `appSwitcherState` service should emit an event on change, and this directive should listen for it. The listener for that event would update `$scope.isAppSwitcherOpen` and do the work of `updateAppSwitcher`. `toggleAppSwitcher` would simply have to call `appSwitcherState.setOpen`. This would be a bit more fluxy.

cjcenizal replied:
You're right, I totally got the data flow going in the wrong direction. Great catch!

Bargs commented on src/ui/public/chrome/directives/kbn_chrome.js:
```diff
@@ -43,6 +43,29 @@ export default function (chrome, internals) {
+          updateAppSwitcher($scope.isAppSwitcherOpen);
+
+          // Notify visualizations, e.g. the dashboard, that they should re-render.
+          $scope.$root.$broadcast('ready:vis');
```
Is this how the visualizations get re-rendered on window re-size? I only see this getting emitted from visualize.js and doc_table.js atm, so I'm a little worried this might be misusing an existing event, but I don't know enough about Visualize to be sure.

cjcenizal replied:
In the `grid`, used for dashboards, you can see that both the window `resize` event and this event are handled by the same layout function: https://github.com/elastic/kibana/blob/master/src/core_plugins/kibana/public/dashboard/directives/grid.js#L102

I tested it out and it has the desired result of re-rendering a dashboard. It doesn't look like it breaks any of the other plugins which listen to it (Discover and Visualize). What do you think? I feel pretty good about using it.

Bargs replied:
I worry a bit about piggy backing on an existing event. Even if it doesn't have unintended side effects today, it might surprise other devs in the future. If I added a `ready:vis` listener I certainly wouldn't expect it to fire on every sidebar expand/collapse.

What about defining a new event that the chrome can fire, and listening for it in the grid.js file you linked to?

cjcenizal replied:
I think you're completely right. I didn't investigate the `ready:vis` event thoroughly enough. Now that I've dug into what's listening to it and how it's being handled, I don't think we want to dispatch it at all. Thanks for explaining your thoughts.

Bargs commented on src/core_plugins/kibana/public/assets/play-circle.svg:
```diff
@@ -0,0 +1,3 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="8" height="8" viewBox="0 0 8 8">
+  <path d="M4 0c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm-1 2l3 2-3 2v-4z" />
+</svg>
```
Since this is used in the chrome across apps, should it be in `src/ui/public/icons`?

cjcenizal replied:
It seems to have more in common with the other icons in `src/core_plugins/kibana/public/assets`. Do you mean to say that the kibana core_plugin is optional, so this icon could possibly be excluded from someone's deployment?

Bargs replied:
Deleting specific plugins was actually a strategy some folks used in the past, though it's less common now that it requires modifications to the code as well. Even though they're "core", I think it's best to avoid having common components rely on assets inside a plugin. It's weird to me that the logout icon is in that folder as well, but the others make sense because they're apps provided by the kibana plugin.

cjcenizal replied:
I tried implementing your suggestion, but our build process doesn't copy those files over to the optimize directory, probably because webpack. Any ideas on a way around this? This problem also blocks https://github.com/elastic/kibana/pull/7911 so if we can solve it, that would be fantastic.

Bargs replied:
It looks to me like they do get copied over, but the files are renamed with a long hash. The image probably needs to be imported so webpack has a path to follow. I'm guessing the existing icons in src/ui/public/icons get included because of this less file https://github.com/Bargs/kibana/blob/46c3eaca6e80859b42964ebaf11f77a7069b233a/src/ui/public/styles/base.less#L640-L640

cjcenizal replied:
I hear you. I'm pretty stumped here. How about we create an issue to move the logout and play-circle icons to the appropriate place, and address this later?

Bargs commented:
@cjcenizal I'm all done reviewing for now. Let me know what you think about the comments above. I've run out of time today, so I'll get to the other x-plugins PR tomorrow.

cjcenizal commented:
@Bargs I've addressed most of your feedback. In terms of naming, I think all of the names we're using right now are confusing, so maybe we can clean things up a bit in this PR.

Currently, kbn_chrome.js uses chrome.html as a template. Let's rename that to kbn_chrome.html.

The sidebar is defined by `<nav class="app-links-wrapper">` -- the idea of "app links" makes very little sense to me. Let's call this component a "Global Nav", which I think reflects its role better and complements the concept of a [Local Nav](https://elastic.github.io/kibana-ui-framework/#/localnav) (@alt74 feel free to chime in).

Let's also encapsulate this markup inside its own directive: `<global-nav>`. We'll need to rename classes and child directives to reflect this new name. `appSwitcherState` will become `globalNavState`.

Thoughts?

Bargs commented:
+:100: to everything you said about naming. The hardest part about reviewing this PR was wrapping my head around the *existing* naming schemes.

cjcenizal commented:
@Bargs did the refactor and ready for your 👀  again

Bargs commented:
jenkins, test this

cjcenizal commented:
jenkins, test this

Bargs commented:
These links added by x-pack at the bottom aren't showing up in this PR branch (screenshot is from master):

<img width="525" alt="screen shot 2016-08-10 at 7 37 33 pm" src="https://cloud.githubusercontent.com/assets/6239176/17574829/3ffa5376-5f32-11e6-9f7a-70eb2b384b51.png">

Bargs commented:
Aside from the issue above, everything looked good. I'll look at it one more time after that's fixed, but you might want to get another set of fresh eyes since a lot has changed since the previous reviewer looked at it. I always get a bit paranoid with big renames.

cjcenizal commented:
@Bargs Thanks! I'll get another pair of eyes. The X-Plugins don't show up because of all the changes in this PR. Once I get a LGTM I will update the X-Plugins PR to be compatible with this one.

Bargs commented:
Ahhhh, right, I totally forgot about that. My brain is burnt out for today I think

spalger commented on src/ui/public/chrome/directives/global_nav/app_switcher/app_switcher.js:
```diff
@@ -72,6 +72,15 @@ uiModules
       // so we force them when needed
       this.ensureNavigation = appSwitcherEnsureNavigation;
+
+      $scope.getTooltip = link => {
```
I would prefer that you stuck with the controller concept here and used `this.getTooltip()` and `switcher.getTooltip()` in the view.

cjcenizal replied:
I'm not opposed but could you explain the rationale for controller over scope methods?

spalger replied:
`$scope` is a perfectly fine way of exposing values to the view, but it has drawbacks when you start involving child scopes (shadowing, unreflected writes, etc.)

Since around angular 1.5 and the new component type, controllers have just been promoted to be the preferred way to expose values to the view as they provide a namespace preventing the inherited values problem, they can be required by child directives/components, and I think they generally describe their purpose better (the controller controls the view, not the `$scope`).

spalger commented:
One note, nothing serious, looks great!

spalger commented:
LGTM
